Add GitHub Actions - Ubuntu, macOS, & Windows, Ruby 2.4 thru head, misc updates

MSP-Greg opened the pull request:
See https://github.com/MSP-Greg/redcloth/actions/runs/7121117282

Some of the changes are 'general cleanup', some are Windows specific changes.

I believe this needs PR #82 merged for Windows CI to pass.

closes #63, closes #71, closes #81

MSP-Greg commented:
As mentioned in #82, this started this morning, and I now actually looked at other PR's and issues.  Added 'co-author' both here and in #82, and also added all the issues and pr's closed by them...

PR #74 should also be merged.  Haven't looked at the others.

MSP-Greg commented:
@digitalmoksha @heliocola

Sorry for the ping.  I forgot that I was a collaborator here.  I am busy enough with other OSS, and I don't know Textile docs.

So, I'm not sure if I can help with real code issues here, but the infrastructure things (build scripts, CI, etc) I can help with.

I think #74, #82, and this PR can be merged, and they'll give a good start to updating things, along with CI.  I haven't worked with any CI going back before Ruby 2.4 for a while.  I know Windows is really a PITA, not sure about macOS or Ubuntu.  I did not add a 
`required_ruby_version` to the gemspec.  Also, deleted the rvm stuff here.

So, wondering if you have any opinions...

heliocola commented:
Hi there @MSP-Greg . Thank you for submitting this PR.
I've just merged #81 that was queued up before yours.

Can you please sync your branch with latest master?

heliocola commented on tasks/rvm.rake:
```diff
@@ -1,81 +0,0 @@
-require 'rvm'
```
@MSP-Greg : can you please revert this change?
I still used this one when I released the last version of the gem.

I want to do a clean up as well, but not in the scope of this PR.

MSP-Greg replied:
From the current file, which was last updated in Apr-2016...
```ruby
RVM_RUBIES = ['ruby-1.8.6-p398', 'ruby-1.9.1-p243', 'ruby-1.9.2-p136', 'ruby-2.2.3p173']
```

> I still used this one when I released the last version of the gem.

For testing?  That's what CI is for.  BTW, releases, maybe a sensible `required_ruby_version` in the gemspec?  I'd suggest 2.4...

heliocola replied:
@MSP-Greg : that is not for testing.

heliocola commented on Gemfile:
```diff
@@ -1,7 +1,4 @@
 source 'https://rubygems.org'
 gemspec
 
-group :compilation do
```
@MSP-Greg : same here. Can you please revert this?
I will clean this up later. I am still situating myself and making sure I situate myself before starting deleting and cleaning up.

heliocola commented on redcloth.gemspec:
```diff
@@ -14,7 +14,6 @@ Gem::Specification.new do |s|
   s.rubyforge_project = "redcloth"
 
   s.rubygems_version   = "1.3.7"
-  s.default_executable = "redcloth"
```
👍 
I've noticed this warning and this was on my TODO list.

heliocola reviewed (changes requested):
@MSP-Greg : thanks again for sending this PR.
Please take a look at my comments.

digitalmoksha commented:
I don't see the actions running in this branch, so I pulled it into a fork. The Windows version is not building properly, but I think that's what #82 / #62 is supposed to fix.

Once this is rebased, I guess we can remove `ruby.yml`

Overall this looks pretty good!

heliocola commented:
> I'll wait until https://github.com/jgarber/redcloth/pull/62 or https://github.com/jgarber/redcloth/pull/82 are merged, as CI will fail until then.

@MSP-Greg : I've merged #62 , can you please check if that suffice for this PR?

I spent some time reading your comments in different issues to situate myself on all the windows issues you've had in the past years (way before I even knew this gem existed).

This comment caught my eye: https://github.com/jgarber/redcloth/issues/51#issuecomment-375977071

It does look like there is a better way to do these extensions and nokogiri is a really good reference. Thank you for pointing that up, I will try to spend some time studying nokogiri's code.

As I've mentioned in a few other comments I am still trying to situate myself so I don't break anything... :-D

digitalmoksha commented:
> It does look like there is a better way to do these extensions and nokogiri is a really good reference. Thank you for pointing that up, I will try to spend some time studying nokogiri's code.

I think the common way of doing this is with https://github.com/rake-compiler/rake-compiler.

MSP-Greg commented:
`rake-compiler` is in the Gemfile and is being used.

MSP-Greg commented:
I just rebased, added Ruby 3.3, and made use of a new feature in setup-ruby.

All passed, see https://github.com/MSP-Greg/redcloth/actions/runs/7590837155/job/20678021321

Obviously, it takes longer than the current CI, but it tests all platforms.  Note there are some warnings thrown in the compile step...

heliocola commented:
Thank you @MSP-Greg .

I will look further in the gem release process and perhaps work on a clean up PR that I've been planning.
I was trying to avoid having to look into that before closing these issues, but perhaps is better that I work on it first.

heliocola commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,63 @@
+      fail-fast: false
+      matrix:
+        os: [ ubuntu-20.04, ubuntu-22.04, macos-12, macos-13, windows-2022 ]
+        ruby: [ 2.4, 2.5, 2.6, 2.7, '3.0', 3.1, 3.2, 3.3, head ]
```
@MSP-Greg : I just synced this with the lastest master branch and it is failing on CI.
Can you please check if I messed up something?

The latest master has the latest PRs merged, one clean up PR I've just pushed, and the release of v4.3.4.

And as far as Ruby versions on CI, I do not want to run all these, just the Ruby releases that are being maintained by the Ruby core team: 3.1 and above (given 3.0 is EOL in a few weeks).

heliocola replied:
FYI: this is what I had in mind: #87

MSP-Greg replied:
I just did a proper rebase, along with removing Ruby 2.3 and earlier from Actions CI.

> And as far as Ruby versions on CI, I do not want to run all these, just the Ruby releases that are being maintained by the Ruby core team: 3.1 and above (given 3.0 is EOL in a few weeks).

You should run CI matching the gemspec, which is:

https://github.com/jgarber/redcloth/blob/8b10f41636616562d822d62382f64cbce24969e4/redcloth.gemspec#L16
